File: services/script-service/src/services/scripts.ts
// Business logic for scripts (1:1 with projects).
// 所有 DB 访问都通过 withTeamContext, 让 RLS 兜底.

import { eq, sql } from 'drizzle-orm';

import { conflict, notFound } from '../apperr.js';
import { withTeamContext } from '../db/client.js';
import { scripts, type Script } from '../db/schema.js';

export interface AuthCtx {
  userId: string;
  teamId: string;
  role: 'owner' | 'admin' | 'editor' | 'viewer';
}

// ---- get or create ----

export async function getScript(ctx: AuthCtx, projectId: string): Promise<Script> {
  return withTeamContext(ctx.teamId, ctx.userId, async (tx) => {
    const rows = await tx
      .select()
      .from(scripts)
      .where(eq(scripts.projectId, projectId))
      .limit(1);
    if (rows.length > 0) return rows[0]!;
    // 自动创建空 script
    const inserted = await tx
      .insert(scripts)
      .values({ projectId, content: '', updatedBy: ctx.userId })
      .returning();
    return inserted[0]!;
  });
}

// ---- update with optimistic lock ----

export interface UpdateScriptInput {
  content: string;
  expectedVersionNo: number;
}

export async function updateScript(ctx: AuthCtx, projectId: string, input: UpdateScriptInput): Promise<Script> {
  return withTeamContext(ctx.teamId, ctx.userId, async (tx) => {
    const rows = await tx
      .select()
      .from(scripts)
      .where(eq(scripts.projectId, projectId))
      .limit(1);

    if (rows.length === 0) throw notFound('脚本不存在');
    const current = rows[0]!;

    if (current.versionNo !== input.expectedVersionNo) {
      throw conflict('版本冲突, 请刷新后重试', {
        current_version_no: current.versionNo,
        expected_version_no: input.expectedVersionNo,
      });
    }

    const wordCount = countWords(input.content);
    const sceneCount = countScenes(input.content);

    const updated = await tx
      .update(scripts)
      .set({
        content: input.content,
        wordCount,
        sceneCount,
        updatedBy: ctx.userId,
        versionNo: sql`${scripts.versionNo} + 1`,
      })
      .where(eq(scripts.projectId, projectId))
      .returning();

    return updated[0]!;
  });
}

// ---- helpers ----

function countWords(content: string): number {
  if (!content) return 0;
  // 中文字符数 + 英文单词数
  const chineseChars = content.match(/[一-鿿]/g);
  const withoutChinese = content.replace(/[一-鿿]/g, ' ');
  const englishWords = withoutChinese.trim().split(/\s+/).filter(Boolean);
  return (chineseChars?.length ?? 0) + englishWords.length;
}

function countScenes(content: string): number {
  if (!content) return 0;
  const lines = content.split('\n');
  return lines.filter((l) => l.startsWith('## ')).length;
}

File: services/ai-gateway/app/repo/scripts.py
"""scripts repo —— ai-gateway 直写共享库(与 shots 同套 team_ctx 写法)。

scripts 表归 script-service,但 ai-gateway 无 S2S HTTP 客户端,故经 team_ctx 直读写,
RLS 兜底(与 repo/shots.py 同样的理由)。单场重写用乐观版本:
UPDATE ... WHERE project_id=$ AND version_no=$expected,版本不符即冲突。
"""

from __future__ import annotations

import uuid as uuidlib

from ..db import team_ctx
from ..scene_split import split_scenes


async def get_script(*, team_id: str, user_id: str, project_id: str) -> dict | None:
    pid = uuidlib.UUID(project_id)
    async with team_ctx(team_id, user_id) as conn:
        row = await conn.fetchrow(
            "SELECT content, version_no FROM scripts WHERE project_id = $1", pid
        )
    if row is None:
        return None
    return {"content": row["content"], "version_no": row["version_no"]}


async def update_script_content(
    *, team_id: str, user_id: str, project_id: str, content: str, expected_version_no: int
) -> int | None:
    """乐观版本写回。返回新 version_no;版本不匹配(冲突)返回 None。

    word_count/scene_count 是派生展示字段,这里随内容一并重算(scene_count 用同一切分契约)。
    """
    pid = uuidlib.UUID(project_id)
    scene_count = len(split_scenes(content))
    word_count = len(content)
    async with team_ctx(team_id, user_id) as conn:
        row = await conn.fetchrow(
            """UPDATE scripts
                  SET content = $1,
                      version_no = version_no + 1,
                      word_count = $2,
                      scene_count = $3,
                      updated_by = $4,
                      updated_at = now()
                WHERE project_id = $5 AND version_no = $6
              RETURNING version_no""",
            content,
            word_count,
            scene_count,
            uuidlib.UUID(user_id),
            pid,
            expected_version_no,
        )
    return row["version_no"] if row else None
